### ISSUE TITLE
Authorization (lock symbol) is rendered incorrectly

### ISSUE THREAD

**Issue by m-mohr** [swagger-api/swagger-ui#4402]
I have endpoints that either have a required authorization or an optional authorization (see example).
I think the lock symbols are rendered incorrectly. It shows a locked symbol for optional authorization (/public) and and an unlocked symbol for required authorization (/private).

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug
| Which Swagger/OpenAPI version?  | 3.0.0 and 2.0
| Which Swagger-UI version?       | 3.x currently used by hosted Swagger Editor + master branch (03.04.2018 15:36)
| How did you install Swagger-UI? | Hosted Swagger Editor + locally using master branch (03.04.2018 15:36)
| Which browser & version?        | Chrome 65.0.3325.181
| Which operating system?         | Windows 10

### Demonstration API definition

```yaml
openapi: 3.0.0
servers:
  - url: 'https://localhost/api/'
info:
  title: OpenEO API
  version: 0.3.0
paths:
  /public:
    get:
      summary: This endpoint allows users to access it with AND without authentication.
      security:
        - {}
        - Bearer: []
      responses:
        '200':
          description: ...
  /private:
    get:
      summary: This endpoint allows users to access it only with authentication.
      security:
        - Bearer: []
      responses:
        '200':
          description: ...
components:
  securitySchemes:
    Bearer:
      type: http
      scheme: bearer
```

### Expected Behavior
It shows an **unlocked** symbol for optional authorization (/public) and and an **locked** symbol for required authorization (/private). Colors are wrong, too.

### Current Behavior
It shows a **locked** symbol for optional authorization (/public) and and an **unlocked** symbol for required authorization (/private). Colors should be swapped, too.

**Comment by webron** [comment]
I understand the confusion, but it's actually working as expected.

When a user fills the authorization, the lock becomes closed and black - that indicates that there's security information provided. An unlocked lock, means that the user has not provided the information. We've had discussions in the past about how some people expect it to be one way and some the other. We'll consider changing it altogether to make it clearer.

In your case, it behaves as expected (in our intent) - since you allow a no-security option, meaning the user can use the call without providing credentials, the lock is black and locked indicating you can execute the call.

**Comment by IceflowRE** [comment]
Thats more than confusing.

First a better example:
```
openapi: 3.0.0
servers:
  - url: 'https://localhost/api/'
info:
  title: OpenEO API
  version: 0.3.0
paths:
  /public:
    get:
      summary: This endpoint allows users to access it only without authentication.
      responses:
        '200':
          description: ...
  /misc:
    get:
      summary: This endpoint allows users to access it with AND without authentication.
      security:
        - {}
        - Bearer: []
      responses:
        '200':
          description: ...
  /private:
    get:
      summary: This endpoint allows users to access it only with authentication.
      security:
        - Bearer: []
      responses:
        '200':
          description: ...
  
components:
  securitySchemes:
    Bearer:
      type: http
      scheme: bearer
```

Three states exist and the natural interpretation of those is easy:

| state | interpretation |
|-|-|
|none| No authentication required. |
|open| Optional authenticaton, thats why its already open, because you can access it even without. |
|locked| You need an authorization to open it and only with you can access it. |

To signalize a provided security information, mark the corresponding locks green or so.

**Comment by RolandoAvila** [comment]
Can someone provide the explanation that clarifies how this is not confusing? I would like to know how to explain it to the consumers to my APIs.

**Comment by shockey** [comment]
@RolandoAvila, @webron summed it up above pretty well IMO.

**Comment by RolandoAvila** [comment]
Thanks @shockey , yes I read his comment, but I guess it's not convincing me that is an intuitive solution. If it was, I shouldn't have to continuously explain this to those new to swagger's documentation (which I love working with btw). I even surveyed some random non-technical people and sure enough all had the same confused conclusion I had.

**Comment by MikeGriffinReborn** [comment]
Very poor design. The API is locked until I authenticate then it is unlocked to me. I suggest you guys change this. It's beyond wondering how the logic behind the current logic came about.

**Comment by anasoftinc** [comment]
This is actually a defect to be fixed since it is showing wrong information about API method on Swagger UI .
It should be unlocked image for API method attributed [AllowAnonymous].

**Comment by coloradocolby** [comment]
Agreed, the logic on this is completely backwards. I'd suggest listening to user feedback on this one as it's clearly a common complaint.

**Comment by FirstChoice-Robert** [comment]
I'll add my voice to this as well. The current implementation is confusing. The open padlock to me symbolises that the end point is unlocked and can be used.The black padlock means that the end point is currently locked and can not be used without authorisation

**Comment by andi-livn** [comment]
I was about to raise this issue but found this existing ticket. 

It may be splitting hair, but I very much agree that the notion of a closed=locked padlock to the common user is that something is secured against unauthorised access and will require a key to gain access, whereas an open=unlocked padlock would normally be interpreted as something that may otherwise be locked, but has already been unlocked.

As such, I too suggest an overhaul of the icons used to represent the different security requirements and Swagger UI session states:

Operations could have:

- No padlock icon when no security has been defined
- An open padlock when authentication is optional or after credentials matching any of the defined security schemes have been provided
- A closed padlock when security is mandatory and no suitable credentials have been given

The Authorize button should use a closed/locked padlock before and an open/unlocked padlock after the user has provided credentials, so essentially the exact reverse of the current behaviour. Alternatively, a crossed-out or grey key icon could be displayed, which turns black once the user has logged in.

**Comment by ericsampson** [comment]
This is nutty. If you picked a random sampling of Swagger consumers and showed them the current behavior, I bet 95 of them would say that it's backwards.
Would a PR be accepted to change this UI?

**Comment by Mohammedalasaad** [comment]
Yes me too.. This is very confusing :)  I agree with @bmbell I think this should be reversed :D ...

> I found the current implementation very confusing. Just take a real-life scenario - if something is locked (in this case, the black lock symbol), then it generally means it cannot be accessed without a key (e.g. an api key). The key "unlocks" the service, and grants access to it. Thus, the open lock symbol would be shown after credentials were entered.

**Comment by JonPugsley** [comment]
Another vote for reversing here - to me displaying a closed padlock implies access to the method is locked/prohibited, an open padlock implies that the method is secured but I have provided the correct "key", and no padlock implies free/public access.

**Comment by jez9999** [comment]
I'd add that in addition to this fix, swagger-ui should detect when the *only* entry in the security array is the empty object, and mark that method as anonymous (no padlock), as I've detailed in #1819.  The reason for this is that with Swashbuckle / OpenAPI.NET, there is no way to actually override an operation's global security and remove it completely to create an empty array - the most that can be done is an array with one empty object in it.  Apparently this is supported, and indicates an anonymous operation - NOT one with optional security.

**Comment by marcotitoL** [comment]
i too found the lock and unlock icon confusing.
im just new and still learning about dotnet and swagger.

what i did is that, i added some js to swap the icons 😂 

```
app.UseSwaggerUI(
swagger => {
    swagger.InjectJavascript("/changelock.js");
}
```

and on that js file is this:

```
window.addEventListener("load", (event) => {
  setTimeout(() => {
    var lockedIcon = document.getElementById("locked");
    var unlockedIcon = document.getElementById("unlocked");
    lockedIcon.id = "unlocked";
    unlockedIcon.id = "locked";
  }, 100);
});
```

**Comment by angelorscoelho** [comment]
> i too found the lock and unlock icon confusing. im just new and still learning about dotnet and swagger.
> 
> what i did is that, i added some js to swap the icons 😂
> 
> ```
> app.UseSwaggerUI(
> swagger => {
>     swagger.InjectJavascript("/changelock.js");
> }
> ```
> 
> and on that js file is this:
> 
> ```
> window.addEventListener("load", (event) => {
>   setTimeout(() => {
>     var lockedIcon = document.getElementById("locked");
>     var unlockedIcon = document.getElementById("unlocked");
>     lockedIcon.id = "unlocked";
>     unlockedIcon.id = "locked";
>   }, 100);
> });
> ```

Don't attempt to make it work this way. It doesn't work and if it did it would be very prone to fail. The swagger scripts don't only rely on class names or Ids to assert the style/action as other metadata is needed. The current locked/unlocked icon logic is broken and needs a fix. You can't argue that a closed lock indicates a successful access over an open one.

**Comment by marcotitoL** [comment]
> Don't attempt to make it work this way. It doesn't work and if it did it would be very prone to fail. The swagger scripts don't only rely on class names or Ids to assert the style/action as other metadata is needed. The current locked/unlocked icon logic is broken and needs a fix. You can't argue that a closed lock indicates a successful access over an open one.

i am in no way suggesting this as a fix :D i was just doing it for my sanity at that time. and yes it works, as the icons are just html tags, you can do anything with html elements. but again, this is not the correct way to fix it.

**Comment by angrocode** [comment]
```js
  const authorizeIcons = (system) => {
    const React = system.React
    return {
      wrapComponents: {
        SvgAssets: (original) => (props) => {
          const originalSvg = original().props.children.props.children.props.children
          const svg1 = React.createElement('symbol', {viewBox: '0 0 512 512', id: 'unlocked-key'},
            React.createElement('path', {d: "M282.3 343.7L248.1 376.1C244.5 381.5 238.4 384 232 384H192V424C192 437.3 181.3 448 168 448H128V488C128 501.3 117.3 512 104 512H24C10.75 512 0 501.3 0 488V408C0 401.6 2.529 395.5 7.029 391L168.3 229.7C162.9 212.8 160 194.7 160 176C160 78.8 238.8 0 336 0C433.2 0 512 78.8 512 176C512 273.2 433.2 352 336 352C317.3 352 299.2 349.1 282.3 343.7zM376 176C398.1 176 416 158.1 416 136C416 113.9 398.1 96 376 96C353.9 96 336 113.9 336 136C336 158.1 353.9 176 376 176z"})
          )
          return React.createElement('div', null,
            React.createElement('svg', {xmlns: 'http://www.w3.org/2000/svg', xmlnsXlink: 'http://www.w3.org/1999/xlink', className: 'svg-assets'},
              React.createElement('defs', null, [svg1, ...originalSvg])
            )
          )
        },
        authorizeBtn: () => (props) => {
          const { isAuthorized, showPopup, onClick, getComponent } = props
          const AuthorizationPopup = getComponent('authorizationPopup', true)
          return React.createElement('div', {className: 'auth-wrapper'},
            [
              React.createElement('button', {className: isAuthorized ? 'btn authorize unlocked' : 'btn authorize locked', onClick},
                [
                  React.createElement('span', null, 'Authorize'),
                  React.createElement('svg', {width:'20', height: '20'},
                    React.createElement('use', {href: isAuthorized ? '#unlocked' : '#locked', xlinkHref: isAuthorized ? '#unlocked' : '#locked'})
                  )
                ]
              )
            ],
            showPopup && React.createElement(AuthorizationPopup)
          )
        },
        authorizeOperationBtn: () => (props) => {
          const { isAuthorized, onClick } = props
          return React.createElement('button', {className: isAuthorized ? 'authorization__btn unlocked' : 'authorization__btn locked', onClick},
            React.createElement('svg', {width:'20', height: '20'},
              React.createElement('use', {href: isAuthorized ? '#unlocked-key' : '#locked', xlinkHref: isAuthorized ? '#unlocked-key' : '#locked'})
            )
          )
        }
      }
    }
  }
```
![image](https://user-images.githubusercontent.com/6058779/172920607-0427d049-cb51-4459-bf81-68dbecc98224.png)

**Comment by fannypwr** [comment]
another vote for changing the current design - I was totally confused

**Comment by johnwc** [comment]
How is this still going on 5 years later. A lock icon to anyone in the world is a symbol of no access without key/secret. The default should be lock icon set on all endpoints with security attribute applied, unlock icon for those that are unsecured with security or has been authenticated with a key in the UI.

If this is such a stickler thing that you do not agree with, then make it an option at least for others to opt in to one way or the other.

**Comment by bashirmja** [comment]
I made a small improvement meanwhile we are waiting to fix this issue.  Did not test it much but works with the latest version at least. 

```JavaScript 
const ui = SwaggerUIBundle({
      url: "/swagger/docs/v1/",
      dom_id: '#swagger-ui',
      onComplete: function () {
         console.log(JSON.stringify(versions));
          var lockedIcon = document.querySelector(".svg-assets defs symbol#locked");
          var unlockedIcon = document.querySelector(".svg-assets defs symbol#unlocked");
          lockedIcon.id = "unlocked";
          unlockedIcon.id = "locked";
      },
});
```

**Comment by maxenceTroislouche** [comment]
I think the current implementation very confusing and should actually be the opposite :/

**Comment by andrea14polanco-sa** [comment]
This is still very confusing. I was about to raise a bug but then found this.

**Comment by bubike** [comment]
ditto

**Comment by huntercavazos** [comment]
Really seems like a no brainer, especially when your user-base requests otherwise. How is this still not fixed?

**Comment by ArtA110** [comment]
There is absolutely no logic in marking public APIs as LOCKED and privates as UNLOCKED.
Please fix this, it is so confusing

**Comment by nzottmann** [comment]
> When a user fills the authorization, the lock becomes closed and black - that indicates that there's security information provided. An unlocked lock, means that the user has not provided the information.

The security information for a padlock is a key.
- If you do not provide the key, the lock is closed
- If you provide the key, the lock is open

For most padlocks, you do not need a key to switch it from open to closed. For all padlocks, you can only switch from closed to open by providing a key.

If you continue using a padlock here, please switch open and locked to match physical padlocks.
